UMAprotocol/snapshot
# feat: add osnap plugin

Changes in this PR:

We have had great success offering oSnap as an optional alternative to Reality in the SafeSnap plugin. However, over time our needs have evolved. We have determined that to offer the best experience to our users, we need our own Snapshot plugin.

This PR seeks to add and integrate this new plugin. 

We have intentionally not yet removed the legacy oSnap logic from the existing SafeSnap plugin, because we want users who have not migrated to still be able to use the old plugin as they have done up until now.

The new features / improvements of this plugin are:

* Simplified transaction builder — since oSnap can execute an array of transactions and doesn't require them to be formatted as batches, we are able to greatly simplify the transaction builder UI.
* Automatic module address detection — users no longer need to specify network and module addresses for safes in the plugin config.
* Strongly typed data — we have types for all inputs and outputs that the plugin uses.

While we have wherever possible avoided making changes outside of the `/plugins` directory, there are some changes that exist outside:

* in `SpaceCreateVoting` we have a `use osnap` checkbox (same as with the SafeSnap plugin)
* in `SettingsTreasuries` we have an `enable osnap` button next to each treasury (only shown if the oSnap plugin is installed)

How to test and review this PR?

To test, we must run through the plugin's features and verify that they work on chain.

1. Enable / disable oSnap

Install the plugin on a space with Safe treasuries.
The `activate osnap` button should be visible.
Clicking the button should take the user to user Safe app.
Returning to Snapshot, the button for that treasury should now say `oSnap active`
Clicking that button again should take you to the Safe app again, where oSnap can be deactivated.

2. Transaction builder (Create)

Create a proposal and choose to use oSnap.
The final step should be the transaction builder.
The user should be able to add an arbitrary number of transactions.
The address fields and to fields should be required.
Once transactions are added and the proposal is submitted, the transactions should be persisted to ipfs.

3. Proposal

Once a proposal is created, the transactions should be visible there in a read-only state.
Once the vote is over:

- if the vote failed, the user should not be able to do anything
- if the vote passed, then the user should be able to submit the proposal with their own bond
- if the proposal is submitted and has not been disputed in the Optimistic Oracle, the user should be able to request execution of the proposal
- once execution is requested, the transactions should be completed on chain
